### DISCUSSION TITLE
Storage Reliability on Filecoin

### DISCUSSION THREAD

**Discussion by SHSRPL** [filecoin-project/FIPs#843] [Enhancements - Technical]
# Storage Reliability on Filecoin

Watch the talk from FIL Dev Summit Singapore [here](https://www.youtube.com/watch?v=XFjaZMkk4PM&ab_channel=FilecoinEcosystem).

## Background

CEL observed a sharp increase in the number of sectors being terminated between July-2022 to March-2023. Therefore, we started a GitHub discussion [#691](https://github.com/filecoin-project/FIPs/discussions/691) to track the issue, and hypothesize  reasons for the sharp increase. 

We further spent some time analyzing the termination fee structure in more detail and presented our [report](https://hackmd.io/@1dR0N2W7SQyZWg7DGB8Vfw/HyKRs2AD3) to the public. The TL:DR from our report was that:

- The present termination fee structure was designed in the context of 6 month sectors. The principle that it follows is to cap the termination fee to half the lifetime earnings of a 180 day sector. Since then, the network has onboarded many sectors with durations up to even 1 year and 1.5 years, meaning that such sectors also pay only 90 days worth of block rewards as a penalty. Thus, the current structured is rendered outdated, more so with FIP-052.
- It can be argued that under the current fee strcuture, incentives to maintain sectors may become ineffective due to decreasing block rewards, and decreasing pledge costs. This can potentially affect the power stability of the network.
- This opens up the question of whether the present termination fee structure should be revised.
- An example of such a change could be to (1) remove the `weighting_factor` in the termination fees calculation, which is currently set to `1/2`, and, (2) scale the `TERMINATION_LIFETIME_CAP` with sector duration commitment according to a function  $f(x) = 2(1-e^{-x})$. Such a change can potentially lead to fewer terminations in the network, while still making sure that the penalty-to-reward ratio for SPs continues to be > 1.

However, on tracking the conversations that took place in our initial GitHub discussion, presenting our findings on the core devs meeting, and dicussing our work individually with different members of the community, it was evident that there were divergent views on changing the termination fee. There were roughly two schools of thought on this issue:
1. Increasing the termination penalty is neccessary in order to disincentivize SPs from leaving the network.
2. Decreasing the termination penalty is the way forward since we need to reduce risks that storage providers take on, and lower the entry threshold for them.

There were a lot of hypotheses that were put forth on what would happen to key performance metrics such as network QAP, daily power termination, daily power being onboarded, etc., if there is a change to the termination fee structure, but what was lacking was a quantitative  approach to testing these hypotheses. 

## Agent Based Modelling

We decided to take an [agent-based-modelling](https://github.com/protocol/filecoin-agent-twin) approach in order to  understand the implication of different termination fee policies on key network KPIs such as - daily network QAP, daily power onboarded, daily power renewed and daily power terminated. This would fill in the gap that existed in the discussions by providing a way to concretely understand the potential implications of different termination fee policies on the network. 

The policies that we tested were:
1. The current termination fee policy
2. The increased termination fee policy that we proposed earlier, and 
3. A zero termination fee policy that most people advocated for in the earlier discussion thread

Agent based models are composed of an environment that implements the cryptoeconomic mechanisms of Filecoin (such as locking, vesting, minting, and the interaction between these aspects of token supply), and agents that interact with the environment and model SP behaviour. More details of ABMs can be found in this [article](https://medium.com/cryptoeconlab/modeling-the-filecoin-economy-part-2-9388a6e26fb9). 

**Note**: We are not making predictions of the network KPIs under different policies. Instead, we test different policies in our simulated environment to find the optimal one, and argue that it will also perform well on mainnet. *Simulated network KPIs can differ from what is observed in reality*. 

We built agents that model three types of SPs in the Filecoin network:
1. **Passively Terminating Agents** that terminate a configured amount of power at randomly specified time-steps of the simulation in order to model terminations due to force majeure, such as private key loss or natural disaster. 
1. **Optimistic Agents**  that do not terminate any power due to confidence in Filecoin’s mission and the promise of a greater future ROI
1. **Adaptive ROI Agents** that forecast the ROI from keeping the sector alive for the entirety of it’s duration and also the ROI from terminating their sectors and selling their hardware at its salvage value. Depending on which pathway yields better returns, these agents decide whether or not to terminate a configured amount of power. 

More details on the SP specification such as how the ROI from the different pathways are computed, and how the agents decide when to onboard and renew power can be found [here](https://docs.google.com/presentation/d/1J_1UonroJJGEH5R2k17IyOKmzTlar-4WOalX9Q1VKk4/edit?usp=sharing). The code that implements the mechanisms and behaviour of these agents can be found [here](will add once PR is merged) and [here](https://github.com/SHSRPL/filecoin-agent-twin/blob/main/agentfil/agents/roi_agent_terminate_waves.py). 

### Results 

Given below are the results when we ran the simulation of a network with an equal distribution of agents that believed that the future FIL price would be $0.1, $1, $10, and $100.

![](https://hackmd.io/_uploads/rJgcli91a.png)

![](https://hackmd.io/_uploads/rJkgZsck6.png)
*Fig 1. A) Line plots that describe the total network power (QAP, RBP) in EiB, total QAP onboarded, total QAP renewed, and total QAP terminated during the course of the simulations under each of the three different policies. B) The summary statistics of the same metrics under the same there policies.*

- What we can infer from the following results is that there is no free lunch - the tradeoff is between QAP growth and Storage Reliability. 
    - With a stricter termination fee penalty, the network terminations significantly drops, but at the price of the daily power onboarded and renewed going down too.
    - With a more lenient penalty, the daily power being onboarded and renewed are at healthy levels, but the daily network terminations still continue to stay high.

## Proposal

In order to navigate the tradeoffs between storage reliability and QAP growth we would like to propose the idea of implementing a different policy for different types of SPs, namely CC SPs and FIL+ SPs.

- **Sustained QAP Growth:** For CC SPs, we propose to continue the existing termination, or potentially even implementing a *reduced termination fee policy*. This would allow risk averse SPs the choice to contribute storage power to the network as CC miners. As such, lowering the barrier to entry would allows such SPs to enter the network and contribute to its consensus power when the economic conditions are favourable, while having an out when the conditions worsen and running a sector becomes unprofitable. This would allow the network to maintain healthy levels of QAP being onboarded and renewed. 
- **Robust Storage Reliability**: It is important to remember here that Filecoin is a decentralized data storage network that aims to store humanity’s most important information. It is critical that clients that choose to store their data on the network get **very strong guarantees** for data storage. Even though there exists avenues such as deal collateral for  clients to safeguard against the risk of loosing their data, there is a minimum amount of guarantee that the protocol needs to provide by default. The current termination fee is arguably insufficient to provide this minimum level of guarantee. Therefore, for FIL+ SPs who decide to commit to storing clients' data, we propose implementing an *increased termination fee* policy, such as the one presented in our earlier [report](https://hackmd.io/@1dR0N2W7SQyZWg7DGB8Vfw/HyKRs2AD3).

**Comment by kernelogic** [comment]
It's been 6 months since the CEL charts were published. Does it need to update it to reflect the current situation? A lot has changed since then (on the macro level). Has there been a continue increase of sectors termination?

**Reply by dcasem** [reply]
> Has there been a continue increase of sectors termination?

We've been slowly terminating sectors and will continue to do so until integrity is restored to the network. If there were any changes to the termination costs, this would likely accelerate our plans.

**Reply by anorth** [reply]
> Has there been a continue increase of sectors termination?

No, March/April 2023 appears to have been a local maximum. This chart from Starboard isn't aggregated the same way as the one CEL published, but doesn't appear to show any increase in long term termination rates.

![Daily Terminations (1)](https://github.com/filecoin-project/FIPs/assets/445306/6c425e41-6be9-454b-a66c-70ab1caaea95)

**Comment by jsoares** [comment]
I found the simulation results _slightly_ surprising, in a good way. My opinion on the topic changed during the initial discussion but, based on the data, it seems that the current policy strikes a reasonable balance between competing priorities and there's no pressing need to change it. 

Re data stored, I'm somewhat in favour of the principle of providing a baseline level of guarantee (to [this point](https://github.com/filecoin-project/FIPs/discussions/691#discussioncomment-5749562) in particular, I note that despite the never-ending debate, there's little robust evidence for negative impact of reasonable minimum wage laws), but I'm unconvinced that this is a real problem or that it warrants action and increased complexity.
